homeassistant/components/alarm_control_panel/group.py

```diff
@@ -0,0 +1,241 @@
+
+    @property
+    def code_format(self):
+        """The code format is customizable."""
+        return self._code_format
+
+    @asyncio.coroutine
+    def async_alarm_disarm(self, code=None):
+        """Send disarm command."""
+        yield from self._async_send_message(SERVICE_ALARM_DISARM,
+                                            code=code)
+
+    @asyncio.coroutine
+    def async_alarm_arm_home(self, code=None):
+        """Send arm home command."""
+        yield from self._async_send_message(SERVICE_ALARM_ARM_HOME,
+                                            code=code)
+
+    @asyncio.coroutine
+    def async_alarm_arm_away(self, code=None):
+        """Send arm away command."""
+        yield from self._async_send_message(SERVICE_ALARM_ARM_AWAY,
+                                            code=code)
+
+    @asyncio.coroutine
+    def async_alarm_arm_night(self, code=None):
+        """Send arm night command."""
+        yield from self._async_send_message(SERVICE_ALARM_ARM_NIGHT,
+                                            code=code)
+
+    @asyncio.coroutine
+    def async_alarm_trigger(self, code=None):
+        """Send alarm trigger command."""
+        yield from self._async_send_message(SERVICE_ALARM_TRIGGER,
+                                            code=code)
+
+    @asyncio.coroutine
+    def async_alarm_arm_custom_bypass(self, code=None):
+        """Send arm custom bypass command."""
+        yield from self._async_send_message(SERVICE_ALARM_ARM_CUSTOM_BYPASS,
+                                            code=code)
+
+    @property
+    def state(self):
+        """Return the state of the device."""
+        # _update_state always behaves as if a transition was in progress.
+        # But if the pre- and post-state match, we're not "pending"!
+        self._update_state()
+        if self._pre_state == self._post_state:
+            self._state = self._post_state
+        else:
+            self._state = STATE_ALARM_PENDING
+        return self._state
+
+    def _update_state(self):
+        """Update the current transition state of the device."""
+        def _common_state(states):
+            # This can happen if the sub-panels have not loaded yet.
+            if len(states) == 0:
+                return STATE_ALARM_DISARMED
+
+            # One triggered sub-panel is enough to trigger the whole panel.
+            if states[STATE_ALARM_TRIGGERED]:
+                return STATE_ALARM_TRIGGERED
+
+            # This only happens if a member does not provide pre or
+            # post_pending_state.  In that case, ensure the pending state
+            # percolates up to device_state_attributes, and the attribute
+            # will be left out in the group as well.
+            if states[STATE_ALARM_PENDING]:
+                return STATE_ALARM_PENDING
+
+            # If all entities have the same state, great.  Otherwise,
+            # we can only summarize the state as "custom bypass".
```

Comment:
Why is this? I'm not following here, sorry.

Reply:
Suppose we follow the example in the group documentation, and add a group alarm that is composed of two subpanels (one per floor). The group alarm is then placed under the "home" tab, while the subpanels would go under the "upstairs" and "downstairs" tabs respectively.

You can go to the "upstairs" tab and only arm the "upstairs" control panel. The result is that you have a custom setup for the whole group, where some environments have had their sensors bypassed. The same holds if you arm the entire home via the group alarm, and then disarm the "upstairs" control panel manually. This is expressed with `return STATE_ALARM_ARMED_CUSTOM_BYPASS` here.

Reply:
Thanks, makes sense.